Wire user command events

zhao-oai opened the pull request:
emit new events from user shell tasks, disable sandboxing for user shell commands, and present user shell commands to model

zhao-oai committed:
feat: add user command event types

zhao-oai committed:
feat: run user shell commands via new events

zhao-oai committed:
rebase and refactor

pakrym-oai commented on codex-rs/core/src/event_mapping.rs:
```diff
@@ -15,6 +15,13 @@ use uuid::Uuid;
+fn is_user_shell_command_prefix(text: &str) -> bool {
+    let trimmed = text.trim_start();
+    let lowered = trimmed.to_ascii_lowercase();
+    lowered.starts_with("<user_shell_command>")
```
let's group formatting and parsing in one place like we do with user instructions.

pakrym-oai commented on codex-rs/core/src/tasks/user_shell.rs:
```diff
@@ -78,34 +86,154 @@ impl SessionTask for UserShellCommandTask {
+            Some(Err(err)) => {
+                error!("user shell command failed: {err:?}");
+                let message = format!("execution error: {err:?}");
+                let exec_output = ExecToolCallOutput {
```
do we need to create this struct only to copy data from it right after?

zhao-oai replied:
i think we do since both `format_exec_output_str` and `format_exec_output_for_model` take in an ExecToolCallOutput

pakrym-oai reviewed (changes requested):
Do we need is_user_command on the delta event?
Can we merge the result message into one message? 
Let's colocate message formatting and parsing.

zhao-oai committed:
remove delta event builder since we removed is user shell command field

zhao-oai committed:
fmt

zhao-oai committed:
grouping formatting & parsing + only sending one message for user shell commands

zhao-oai committed:
or_cancel

zhao-oai commented on codex-rs/core/src/tasks/user_shell.rs:
```diff
@@ -78,34 +86,154 @@ impl SessionTask for UserShellCommandTask {
+        tokio::pin!(exec_future);
+
+        let exec_result = tokio::select! {
+            res = &mut exec_future => Some(res),
```
done! using or_cancel() now

pakrym-oai commented on codex-rs/core/src/user_shell_command.rs:
```diff
@@ -0,0 +1,61 @@
+
+pub fn format_user_shell_command_record(command: &str, output: &str) -> String {
+    format!(
+        "{USER_SHELL_COMMAND_OPEN}\n{command}\n{USER_SHELL_COMMAND_CLOSE}\n{USER_SHELL_COMMAND_OUTPUT_OPEN}\n{output}\n{USER_SHELL_COMMAND_OUTPUT_CLOSE}"
```
Consider having only the outer tag and then freeform formatting that matches shell/unified exec inside.

```
Command: ...
Wall clock: ...
Output:
```
etc.

zhao-oai replied:
resolved -- using new freeform output format now

pakrym-oai reviewed (approved):
Few more comments about the exact message content

zhao-oai committed:
remove extraneous check

zhao-oai commented on codex-rs/core/src/event_mapping.rs:
```diff
@@ -42,7 +43,7 @@ fn parse_user_message(message: &[ContentItem]) -> Option<UserMessageItem> {
             }
             ContentItem::OutputText { text } => {
-                if is_session_prefix(text) {
+                if is_session_prefix(text) || is_user_shell_command_text(text) {
```
very true -- sorry i missed this!

zhao-oai committed:
new user shell formatting

zhao-oai committed:
clean up testing

zhao-oai committed:
user shell refactor testing (#6248)

user shell command integration test + unit test ensuring event mapping
ignores user shell commands
